scenario/listboxselect.go

```diff
@@ -0,0 +1,178 @@
+		genericObj := obj.EnigmaObject.(*enigma.GenericObject)
+		settings.doSelect(sessionState, actionState, genericObj)
+	default:
+		actionState.AddErrors(errors.Errorf("Expected generic object , got object type<%T>", t))
+		return
+	}
+
+	sessionState.Wait(actionState)
+}
+
+func (settings ListBoxSelectSettings) doSelect(sessionState *session.State, actionState *action.State, genericObj *enigma.GenericObject) {
+	objInstance := sessionState.GetObjectHandlerInstance(genericObj.GenericId, genericObj.GenericType)
+
+	selectPath, selectType, dataDefType, err := objInstance.GetObjectDefinition(genericObj.GenericType)
+	if err != nil {
+		actionState.AddErrors(err)
+		return
+	}
+	if selectType != senseobjdef.SelectTypeListObjectValues {
+		actionState.AddErrors(errors.Errorf("Object id<%s> type<%s> selecttype<%s> datadeftype<%s> must have selecttype<%s> in listboxselect-action",
+			genericObj.GenericId, genericObj.Type, selectType, dataDefType, senseobjdef.SelectTypeListObjectValues))
+		return
+	}
+
+	var selectFunction func(context.Context, string, bool) (bool, error)
+	switch settings.Type {
+	case All:
+		selectFunction = genericObj.SelectListObjectAll
+	case Alternative:
+		selectFunction = genericObj.SelectListObjectAlternative
+	case Excluded:
+		selectFunction = genericObj.SelectListObjectExcluded
+	case Possible:
+		selectFunction = genericObj.SelectListObjectPossible
+	default:
+		actionState.AddErrors(errors.Wrapf(err, "Unknown filter pane selection type ListBoxSelectType<%s>", settings.Type))
+		return
+	}
+
+	selectListObjects := func(ctx context.Context) error {
+		sessionState.LogEntry.LogDebugf("Select in object<%s> h<%d> type<%s>", genericObj.GenericId, genericObj.Handle, genericObj.GenericType)
+
+		success, err := selectFunction(ctx, selectPath, false)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to select in object<%s>", genericObj.GenericId)
+		}
+		if !success {
+			return errors.Errorf("Select in object<%s> unsuccessful", genericObj.GenericId)
+		}
+
+		sessionState.LogEntry.LogDebug(fmt.Sprint("Successful select in", genericObj.GenericId))
+
+		return err
+	}
+
+	if settings.Wrap {
+		beginSelections := func(ctx context.Context) error {
+			return genericObj.BeginSelections(ctx, []string{selectPath})
+		}
+		sessionState.QueueRequest(beginSelections, actionState, false, fmt.Sprintf("Failed to select in %s", genericObj.GenericId))
+		sessionState.Wait(actionState)
+		if actionState.Errors() != nil {
+			return
+		}
+	}
+	sessionState.QueueRequest(selectListObjects, actionState, false, fmt.Sprintf("Failed to select in %s", genericObj.GenericId))
+	if settings.Wrap {
+		endSelections := func(ctx context.Context) error {
+			return genericObj.EndSelections(ctx, settings.Accept)
+		}
+
+		sessionState.Wait(actionState)
+		if actionState.Errors() != nil {
+			return
+		}
+
+		sessionState.QueueRequest(endSelections, actionState, true, fmt.Sprintf("Failed to end selection in %s", genericObj.GenericId))
+	}
+}
```

Comment:
you should end the action with a `sessionState.Wait`, as it is now, if wrap=false (default) it will exit the action before the selectListObjects request is finished.

Reply:
The caller was waiting, but changing to sync anyways.

Reply:
can't see anything was waiting for sessionState.QueueRequest(selectListObjects, actionState, false, fmt.Sprintf("Failed to select in %s", genericObj.GenericId))

independently, it's not only the request itself we need to wait for, each request response _might_ contain a list of changed objects, requests to update data for these objects will be automatically triggered async int he background. So to wait for there we need to end with a Wait.

Reply:
Alright! There was a third wait last in Execute function at line 107. If this does not wait for sessionState.QueueRequest(selectListObjects, actionState, false, fmt.Sprintf("Failed to select in %s", genericObj.GenericId)) in the case of wrap==false I am misunderstanding something.

Reply:
ah sorry github diff confusing me, for some reason thought this was end of Execution, should have checked it out

Reply:
Don't see anything there.

Reply:
so should still end the action with a Wait for everything that wasn't triggered by "us" directly.

Reply:
Okay, that I did no know, so I removed it. Will put the ending Wait in place again.

Can we be sure that triggered events will not affect the state for the next request, such that Wait in between request are not needed?

Reply:
we don't need to wait for them in this case no, the "reaction" request are fetching layout and data for the objects on the sheet. If we send multiple request effecting the state of an object the engine will send an "aborted" answer to the data request which was ongoing, then the mechanism we have for reaction will re-send it again. Thus getting updated state after the second request.

Reply:
none of the background request will do changes to the actual state in engine, they will however trigger the data to be calculated for the current state.

Reply:
Ahh, I'm learning ;)